Allow writing product attributes via API

alcaeus opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | related to #7811  |
| License         | MIT |

Similar to #7811, but for product attributes.

alcaeus commented on tests/Responses/Expected/product_attribute/show_response_after_partial_update.json:
```diff
@@ -0,0 +1,31 @@
+    "createdAt": "@string@.isDateTime()",
+    "updatedAt": "@string@.isDateTime()",
+    "translations": {
+        "de": {
```
I'm not sure if this is expected behavior: the test updates the attribute and writes new values for the `de_CH` and `en_US` translations. I'd expect it to replace existing translations, not add to the list. What do you think?

tuka217 replied:
The behavior is correct. If you want to see that, the translation for `de_CH` was hanged go to `product_attribute.yml` and set there the translation for `de_CH` (because now it has the translation for `de`).

tuka217 replied:
The behavior is correct. If you want to see that, the translation for `de_CH` was hanged go to `product_attribute.yml` and set there the translation for `de_CH` (because now it has the translation for `de`).

alcaeus commented on tests/Controller/ProductAttributeApiTest.php:
```diff
@@ -85,6 +93,196 @@ public function it_allows_showing_product_attribute()
+}
+EOT;
+
+        $this->client->request('POST', '/api/v1/product-attributes', [], [], static::$authorizedHeaderWithContentType, $data);
```
The create route for product options has a trailing slash - what would you prefer? With or without a trailing slash?

tuka217 replied:
IMO it should be similar to zone's create endpoint. The URL for creating product attribute should have in the path the type of attribute. Then you will not have internal server error while try to call this URL without passing a type.

alcaeus replied:
I can change it - was mainly asking because it was explicitly requested to have the type in the payload when it was last added: https://github.com/Sylius/Sylius/pull/7398#issuecomment-276912669.

tuka217 commented on tests/Controller/ProductAttributeApiTest.php:
```diff
@@ -85,6 +93,196 @@ public function it_allows_showing_product_attribute()
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+
+        $data = <<<EOT
```
`<<EOT` should be moved to the next line

tuka217 commented on tests/Controller/ProductAttributeApiTest.php:
```diff
@@ -85,6 +93,196 @@ public function it_allows_showing_product_attribute()
+        $productAttributes = $this->loadFixturesFromFile('resources/product_attributes.yml');
+        $productAttribute = $productAttributes['productAttribute1'];
+
+        $data = <<<EOT
```
Same here

tuka217 commented on tests/Controller/ProductAttributeApiTest.php:
```diff
@@ -85,6 +93,196 @@ public function it_allows_showing_product_attribute()
+    /**
+     * @test
+     */
+    public function it_allows_delete_to_product_attribute()
```
`allows_to_delete`

tuka217 commented on tests/Controller/ProductAttributeApiTest.php:
```diff
@@ -85,6 +93,196 @@ public function it_allows_showing_product_attribute()
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+
+        $data = <<<EOT
+{
```
And this block should be in one column with other code from this test.

alcaeus commented:
@tuka217 fixed the indentation, also added a second commit to fix heredoc indentation in the code I added in #7811.

As for the translation, I'm not sure I follow. I'd expect the `de` translation to be removed. If this is expected behavior, how do I remove an existing translation for an attribute?

tuka217 commented:
Hi, before I didn't understand you, sorry. You cannot delete the created translation. 
Not sure, but maybe it should be possible during updating product with `PUT`.

tuka217 commented:
And what do you think about this little change for create routing:
```
sylius_admin_api_product_attribute_create:
     path: /product-attributes/{type}
     methods: [POST]
     defaults:
         _controller: sylius.controller.product_attribute:createAction
         _sylius:
             serialization_version: $version
             factory:
                 method: createTyped
                 arguments:
                     type: $type
```
? :)

alcaeus commented:
@tuka217 changed the routing. I'll check out the translation separately, the behavior is the same when PUTting or PATCHing the attribute, so I suspect the form itself.

tuka217 commented:
I meant, that the validation for translation should be fixed and the deletion of not required translation should be possible during `PUT` (but now it is not). But I'm not sure, that it will be a proper solution. For me, it will be ok. 
@michalmarcinkowski

tuka217 commented:
@alcaeus Could you update documentation for api product attributes? :)

alcaeus commented:
@tuka217 added documentation for the product attributes API as well as the product options API from #7811. Let me know if you need anything else 👍

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,142 @@ If you request for more detailed data, you will receive an object with the follo
+Creating a Product Attribute
+----------------------------
+
+To create a new product attribute you will need to call the ``/api/v1/products-attributes/`` endpoint with the ``POST`` method.
```
`api/v1/products-attributes/{type}`

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,142 @@ If you request for more detailed data, you will receive an object with the follo
+Example
+^^^^^^^
+
+To create a new product attribute use the below method:
```
a new **text** product attribute

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,142 @@ If you request for more detailed data, you will receive an object with the follo
+.. warning::
+
+    If you try to create a product attribute without code you will receive a ``400 Bad Request`` error, that will contain validation errors.
+
```
Could you add here the example of a response with validation errors?

alcaeus replied:
Done!

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -463,3 +599,135 @@ Exemplary Response
+                    }
+                }
+            }
+    '
```
Should be in one line with `--data`

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,142 @@ If you request for more detailed data, you will receive an object with the follo
+            {
+                "code": "mug_material"
+            }
+    '
```
Wrong indent

tuka217 commented on docs/api/product_options.rst:
```diff
@@ -20,23 +20,221 @@ If you request a product option via API, you will receive an object with the fol
+            {
+                "code": "MUG_SIZE"
+            }
+    '
```
Wrong indent

tuka217 commented on docs/api/product_options.rst:
```diff
@@ -20,23 +20,221 @@ If you request a product option via API, you will receive an object with the fol
+
+.. warning::
+
+    If you try to create a product option without code you will receive a ``400 Bad Request`` error, that will contain validation errors.
```
Same here. Some example with validation error will be great.

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,185 @@ If you request for more detailed data, you will receive an object with the follo
+        -H "Authorization: Bearer SampleToken" \
+        -H "Content-Type: application/json" \
+        -X POST \
+        --data '
```
`--data(..)` Can be removed.

alcaeus replied:
I'm not sure I follow - did you mean to comment on the empty `--data` argument in the errored call below (removed in the meantime)?

tuka217 replied:
Yea, exactly :D

tuka217 commented on docs/api/product_options.rst:
```diff
@@ -375,3 +608,135 @@ Exemplary Response
+                    }
+                }
+            }
+    '
```
Wrong indent

tuka217 commented on docs/api/product_attributes.rst:
```diff
@@ -48,6 +48,181 @@ If you request for more detailed data, you will receive an object with the follo
+    $ curl http://demo.sylius.org/api/v1/product-attributes/text \
+        -H "Authorization: Bearer SampleToken" \
+        -H "Content-Type: application/json" \
+        -X POST \
```
`\` Should be removed

tuka217 commented on docs/api/product_options.rst:
```diff
@@ -20,23 +20,299 @@ If you request a product option via API, you will receive an object with the fol
+    $ curl http://demo.sylius.org/api/v1/product-options/ \
+        -H "Authorization: Bearer SampleToken" \
+        -H "Content-Type: application/json" \
+        -X POST \
```
`\` should be removed

pjedrzejewski commented:
Awesome work, thanks Andreas! 👍 👍 👍  Just in time for the release. ;)

pjedrzejewski commented:
And thank you Ania for the review! ;)

alcaeus commented:
Thanks Paweł and Ania! Want me to submit another PR to fix the two minor issues in the documentation?
